pennylane/collections/metric_tensor.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains an implementation of the metric tensor using QNode collections.
+"""
+# pylint: disable=protected-access,import-outside-toplevel
+
+import itertools
+
+import numpy as np
+from scipy.linalg import block_diag
+import pennylane as qml
+
+from . import functions as fn
+
+
+def _marginal_prob(prob, wires, interface):
+    """Compute the marginal probability.
+
+    Args:
+        prob (array[float] or tensor[float]): probability vector of size ``(2**num_wires,)``
+        wires (list[int]): list of integers for which to calculate the marginal
+            probability distribution
+        interface (str): the classical interface to target
+
+    Returns:
+        array[float] or tensor[float]: the marginal probabilities, of
+        size ``(2**len(wires),)``
+    """
+    prob = fn[interface].flatten(prob)
+    num_wires = int(np.log2(len(prob)))
+    inactive_wires = list(set(range(num_wires)) - set(wires))
+    prob = fn[interface].reshape(prob, [2] * num_wires)
+    prob = fn[interface].reduce_sum(prob, inactive_wires)
+    return fn[interface].flatten(prob)
+
+
+class MetricTensor:
+    """Construct a QNode collection that computes the metric tensor of a given
+    QNode.
+
+    Args:
+        qnode (QNode): a qubit QNode for which the metric tensor should be computed
+        device (Device, Sequence[Device]): Corresponding device(s) where the resulting
+            metric tensor should be executed. This can either be a single device, or a list
+            of devices of length corresponding to the number of parametric layers in ``qnode``.
+
+
+    **Example**
+
+    Consider the following QNode:
+
+    .. code-block:: python
+
+        @qml.qnode(dev, interface="tf")
+        def qnode(params):
+            qml.templates.StronglyEntanglingLayers(params, wires=[0, 1, 2])
+            return qml.expval(qml.PauliY(0) @ qml.PauliZ(1))
+
+    By using ``qml.MetricTensor()``, we create a metric tensor object that can be evaluated
+    to compute the metric tensor of the QNode:
+
+    >>> g = qml.MetricTensor(qnode, dev)
```

Comment:
Would it make more sense to bind `dev` to a `QNode` instance (not a circuit named "qnode"), and just have `qml.MetricTensor(qnode)`?

Reply:
Have made the change :+1: You can now call

```python
>>> g = qml.MetricTensor(qnode)
```

and `g` will reuse the same device that was bound to `qnode`.

You can _optionally_ pass devices if you wish for `g` to use a different device to `qnode`,

```python
>>> # qnode has three parametrized layers
>>> g = qml.MetricTensor(qnode, devices=[dev1, dev2, dev3])
>>> params = tf.Variable([0.1, 0.2])
>>> g(params, parallel=True)
```

but this is a temporary solution, and we should plan on having a better approach for changing a QNodes device.

Reply:
Nice!

I feel like distributing over multiple devices should be something that is handled by a different function/feature than simply passing a list of 3 devices

Reply:
Yep, that would certainly lead to cleaner code in the collections module (since currently device distribution is left to every function)

Reply:
I've got a lot of thoughts about how to polish up the collections module :grin: